File: native/HevyAlt/components/ui/TextFiled/PasswordValidator.swift
import SwiftUI
import Combine

struct ValidationRequirement: Identifiable {
    var id = UUID()
    let name: String
    let validator: (String) -> Bool
    var isValid: Bool = false
}

class PasswordValidator: ObservableObject {
    @Published var password: String = ""
    @Published var isPasswordVisible: Bool = false
    @Published var requirements: [ValidationRequirement]
    @Published var isValid: Bool = false
    
    private var cancellables = Set<AnyCancellable>()
    
    init(requirements: [ValidationRequirement] = PasswordValidator.defaultRequirements) {
        self.requirements = requirements
        
        $password
            .sink { [weak self] password in
                self?.validatePassword(password)
            }
            .store(in: &cancellables)
    }
    
    private func validatePassword(_ password: String) {
        for i in 0..<requirements.count {
            requirements[i].isValid = requirements[i].validator(password)
            if !requirements[i].isValid {
                for j in (i + 1)..<requirements.count {
                    requirements[j].isValid = false
                }
                break
            }
        }
        
        isValid = requirements.allSatisfy { $0.isValid }
    }
    
    static var defaultRequirements: [ValidationRequirement] = [
        ValidationRequirement(
            name: "8 characters",
            validator: { $0.count >= 8 }
        ),
        ValidationRequirement(
            name: "Number",
            validator: { $0.contains(where: { $0.isNumber }) }
        ),
        ValidationRequirement(
            name: "Special character",
            validator: {
                let specialCharacters = CharacterSet(charactersIn: "!@#$%^&*()_-+=<>?/")
                return $0.unicodeScalars.contains(where: { specialCharacters.contains($0) })
            }
        )
    ]
}

File: native/lib/widgets/text_filed/password_validator.dart
import 'package:flutter/material.dart';

class ValidationRequirement {
  final String name;
  final bool Function(String) validator;
  bool isValid;

  ValidationRequirement({
    required this.name,
    required this.validator,
    this.isValid = false,
  });
}

class PasswordValidator extends ChangeNotifier {
  String password = "";
  bool isPasswordVisible = false;
  late List<ValidationRequirement> requirements;
  bool isValid = false;

  PasswordValidator({List<ValidationRequirement>? customRequirements}) {
    requirements = customRequirements ?? defaultRequirements;
  }

  void updatePassword(String newPassword) {
    password = newPassword;
    _validatePassword(newPassword);
    notifyListeners();
  }

  void toggleVisibility() {
    isPasswordVisible = !isPasswordVisible;
    notifyListeners();
  }

  void _validatePassword(String password) {
    for (var i = 0; i < requirements.length; i++) {
      requirements[i].isValid = requirements[i].validator(password);
      if (!requirements[i].isValid) {
        for (var j = i + 1; j < requirements.length; j++) {
          requirements[j].isValid = false;
        }
        break;
      }
    }
    isValid = requirements.every((r) => r.isValid);
  }

  static List<ValidationRequirement> defaultRequirements = [
    ValidationRequirement(
      name: "8 characters",
      validator: (p) => p.length >= 8,
    ),
    ValidationRequirement(
      name: "Number",
      validator: (p) => p.contains(RegExp(r'[0-9]')),
    ),
    ValidationRequirement(
      name: "Special character",
      validator: (p) => p.contains(RegExp(r'[!@#\$%^&*()_\-+=<>?/]')),
    ),
  ];
}
